app/services/matched_content_change_generation_service.rb

```diff
@@ -10,9 +10,19 @@ def self.call(*args)
   def call
     # if we have records already, then we expect the process completed
     # successfully previously since the insert is an atomic operation
-    return if MatchedContentChange.exists?(content_change: content_change)
+    return if MatchedContentChange.exists?(content_change: content_change) || subscriber_lists.empty?
+
+    now = Time.zone.now
+    records = subscriber_lists.map do |list|
+      {
+        content_change_id: content_change.id,
+        subscriber_list_id: list.id,
+        created_at: now,
+        updated_at: now,
+      }
+    end
 
-    MatchedContentChange.import!(columns, records)
+    MatchedContentChange.insert_all(records)
```

Comment:
Sure, so this table is effective a join table between subscriber lists and content changes. So if we have two processes both trying to add to this table at around the same time they'd be adding the same joins and if they hit a duplicate then it's not likely a problem as they're just adding the same data. 

I put this in as it's an edge case that it's possible that you could have the database change after the MatchedContentChange runs and before this inserts which could then cause a crash. Thinking about it now it might be better to just crash, this would probably be more fitting with the approach we're talking about of applying an advisory lock to prevent concurrent running and is simpler than being ultra idempotent.

I previously had some tests for this on another PR today but decided they were a bit weird:

```
    it "copes in a weird concurrent situation where a different thread inserted "\
       "the matched content changes after our exist query" do
      MatchedContentChange.create!(content_change: content_change,
                                   subscriber_list: subscriber_list)
      allow(MatchedContentChange).to receive(:exists?).and_return(false)

      expect { described_class.call(content_change) }
        .to_not(change { MatchedContentChange.count })
    end
```

Reply:
> if we have two processes both trying to add to this table

I'm still confused, I'm afraid. Could you help me understand the scenario when we would be trying to match the same content change at the same time? I believe this is when we'd expect a duplicate key error.

Reply:
Sure, this is already replaced by the way so somewhat moot, but happy to explain.

Right now, it is possible (ie there is nothing to prevent) two process content change jobs to run at the same time.

If that does happens we probably don't want this to crash on something that isn't actually a critical problem. My expectation is that since both workers would create the same list of MatchedContentChanges then raising an error on duplicate key is an unnecessary crash because it'd just be the same data trying to be re-added.

I've come around to reason in this depth about it though is overly complex and it's probably better to just crash as the dual running isn't something the system was designed for.